### PR TITLE
feat: show all users that have access to a space and their role

### PR DESCRIPTION
**Pull request by Shifan-Gu**
Closes: https://github.com/lightdash/lightdash/issues/8655

### Description:

https://github.com/lightdash/lightdash/assets/101873365/6fcd1e25-2f9e-4254-8360-5ca7c885f63e

I've tested a few scenarios i can think of like create space with permission, switch spaces and manage permissions etc. all seems fine

only caveat is i saw when we witch the access option from "Restrict access" to "Full access" then it will clear all existing restrict access configuration, but i noticed this is an existing behaviour in current released code, so i will leave this to the team to decide how to handle this.

https://github.com/lightdash/lightdash/assets/101873365/38934cce-769d-4bbd-bd80-052dc24d02e0

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

### CODE REVIEW

**Comment by Shifan-Gu** [comment]
@ZeRego thanks for the feedback! i will work on these comments

File: packages/backend/src/models/SpaceModel.ts [hunk 1]
```diff
@@ -397,18 +398,21 @@ export class SpaceModel {
...
                 `users.last_name`,
-
+                // `(case when ${SpaceShareTableName}.user_id is not null then true else false end) as has_direct_access`,
+                `${SpaceShareTableName}.user_id as has_direct_access`,
                 `${ProjectMembershipsTableName}.role as project_role`,
                 `${OrganizationMembershipsTableName}.role as organization_role`,
             ])
-            .distinctOn(`users.user_uuid`)
-            .where(`${SpaceTableName}.space_uuid`, spaceUuid);
+            .where(`${SpaceTableName}.space_uuid`, 'IS', null)
```
[the remark this answers is not in the document]
Comment (Shifan-Gu) [reply packages/backend/src/models/SpaceModel.ts]: done

File: packages/frontend/src/components/common/ShareSpaceModal/index.tsx [hunk 2]
```diff
@@ -91,14 +91,13 @@ const ShareSpaceModal: FC<ShareSpaceProps> = ({ space, projectUuid }) => {
...
-                                space={space}
-                                sessionUser={sessionUser.data}
-                                organizationUsers={organizationUsers}
-                            />
-                        )}
+                        <ShareSpaceUserList
+                            projectUuid={projectUuid}
+                            space={space}
+                            sessionUser={sessionUser.data}
+                            organizationUsers={organizationUsers}
```
[the remark this answers is not in the document]
Comment (Shifan-Gu) [reply packages/frontend/src/components/common/ShareSpaceModal/index.tsx:98]: i will fix this in a separate ticket

**Comment by Shifan-Gu** [comment]
@ZeRego I've tested manually manually myself, let me know if you seeing any issue.

**Review by ZeRego — APPROVED** [review]
lgtm 🎉  Just running e2e tests before merging.
Just had to make a few fixes. Let me know if you don't understand why.

**Comment by Shifan-Gu** [comment]
> lgtm 🎉 Just running e2e tests before merging. Just had to make a few fixes. Let me know if you don't understand why.

Yep, fixes make sense to me, thank you @ZeRego !

**Pull request merged**